sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/FieldBuilder.java

```diff
@@ -0,0 +1,259 @@
+     * @param <T> The generic type of the model class.
+     * @return A collection of fields.
+     */
+    public static <T> List<Field> build(Class<T> modelClass) {
+        ClientLogger logger = new ClientLogger(FieldBuilder.class);
+        return build(modelClass, new Stack<>(), logger);
+    }
+
+    /**
+     * Recursive class to build complex data type.
+     *
+     * @param curClass Current class to be built.
+     * @param classChain A class chain from {@code modelClass} to prior of {@code curClass}.
+     * @param logger {@link ClientLogger}.
+     * @return A list of {@link Field} that curClass is built to.
+     */
+    private static List<Field> build(Class<?> curClass, Stack<Class<?>> classChain, ClientLogger logger) {
+        if (classChain.contains(curClass)) {
+            logger.warning(String.format("There is circular dependencies %s, %s", classChain, curClass));
+            return null;
+        }
+        if (classChain.size() > MAX_DEPTH) {
+            throw logger.logExceptionAsError(new RuntimeException(
+                "The dependency graph is too deep. Please review your schema."));
+        }
+        classChain.push(curClass);
+        List<java.lang.reflect.Field> classFields = Arrays.asList(curClass.getDeclaredFields());
+        List<Field> searchFields = classFields.stream()
+            .filter(classField -> !classField.isAnnotationPresent(FieldIgnore.class))
+            .map(classField -> buildField(classField, classChain, logger))
+            .collect(Collectors.toList());
+        classChain.pop();
+        return searchFields;
+    }
+
+    private static Field buildField(java.lang.reflect.Field classField, Stack<Class<?>> classChain,
+        ClientLogger logger) {
+        Type type = classField.getGenericType();
+
+        if (isSupportedNoneParameterizedType(type)) {
+            return buildNoneParameterizedType(classField, logger);
+        }
+        if (isArrayOrList(type)) {
+            return buildCollectionField(classField, classChain, logger);
+        }
+        List<Field> childFields = build((Class<?>) type, classChain, logger);
+        Field searchField = convertToBasicSearchField(classField, logger);
+        searchField.setFields(childFields);
+        return searchField;
+    }
+
+    private static boolean isSupportedNoneParameterizedType(Type type) {
+        return SUPPORTED_NONE_PARAMETERIZED_TYPE.containsKey(type);
+    }
+
+    private static Field buildNoneParameterizedType(java.lang.reflect.Field classField,
+        ClientLogger logger) {
+        Field searchField = convertToBasicSearchField(classField, logger);
+        return enrichWithAnnotation(searchField, classField, logger);
+    }
+
+
+    private static boolean isArrayOrList(Type type) {
+        if (type.getClass().isArray()) {
+            return true;
+        }
+        return isList(type);
+    }
+
+    private static boolean isList(Type type) {
```

Comment:
Should this check `Collection` as that is the super type for `List` and will capture `Set`, `Queue`, etc.

Reply:
@brjohnstmsft 

What collection type should we support for `Collection (DataType) ` in search? Does service support `Set`, `Queue` etc? I currently only support user to have array or list as collection type.

Reply:
@sima-zhu This seems more like a question about what's idiomatic for Java. As long as it serializes to a JSON array, the service can work with it.

Reply:
Based on service side only take JSON array, I would leave array/list as the choice of collection DataType. 
First, they are enough for most of the use case.
Moreover, as java sdk is not the only entry point of search service, I don't think it is necessary to leave customer the flexibility of `set`, `queue` to manipulate the data which service cannot maintain.